include/sundials/sundials_classview.hpp

```diff
@@ -0,0 +1,120 @@
+/* -----------------------------------------------------------------------------
+ * Programmer(s): Cody J. Balos @ LLNL
+ * -----------------------------------------------------------------------------
+ * SUNDIALS Copyright Start
+ * Copyright (c) 2025, Lawrence Livermore National Security,
+ * University of Maryland Baltimore County, and the SUNDIALS contributors.
+ * Copyright (c) 2013-2025, Lawrence Livermore National Security
+ * and Southern Methodist University.
+ * Copyright (c) 2002-2013, Lawrence Livermore National Security.
+ * All rights reserved.
+ *
+ * See the top-level LICENSE and NOTICE files for details.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ * SUNDIALS Copyright End
+ * -----------------------------------------------------------------------------
+ * Base classes for C++ implementations wrappers (views) of SUNDIALS objects.
+ * ---------------------------------------------------------------------------*/
+
+#ifndef _SUNDIALS_CLASSVIEW_HPP
+#define _SUNDIALS_CLASSVIEW_HPP
+
+#include <memory>
+#include <type_traits>
+#include <utility>
+
+#include <sundials/sundials_convertibleto.hpp>
+
+namespace sundials {
+namespace experimental {
+
+// template<class T, class Creator, class Deleter, class... Args>
+// std::shared_ptr<T> our_make_shared(Args&&... args)
+// {
+//   return std::shared_ptr<T>(Creator{}(std::forward<Args>(args)...), Deleter{});
+// }
+
+template<class T, class Deleter>
+std::shared_ptr<T> our_make_shared(T* ptr)
+{
+  return std::shared_ptr<T>(ptr, Deleter{});
+}
+
+template<class T, class Deleter>
+class ClassView : public sundials::ConvertibleTo<T>
+{
+public:
+  ClassView() : object_(nullptr, Deleter{}) {}
+
+  ClassView(T& object) : object_(std::forward<T>(object), Deleter{}) {}
+
+  ClassView(T&& object) : object_(std::forward<T>(object), Deleter{}) {}
+
+  ClassView(const ClassView&) = delete;
+
+  ClassView(ClassView&& other) { this->object_ = std::move(other.object_); };
+
+  ClassView& operator=(const ClassView&) = delete;
+
+  ClassView& operator=(ClassView&& rhs) = default;
+
+  ~ClassView() { object_.reset(); };
+
+  // Override ConvertibleTo functions
+  T get() override { return object_.get(); }
+
+  T get() const override { return object_.get(); }
+
+  operator T() override { return object_.get(); }
+
+  operator T() const override { return object_.get(); }
+
+protected:
+  std::shared_ptr<std::remove_pointer_t<T>> object_;
+};
+
+template<class Deleter>
+class ClassView<void*, Deleter> : public sundials::ConvertibleTo<void*>
```

Comment:
Is this specialization really necessary?

Reply:
Yes, because we cannot have `shared_ptr<void>` as the `T object` for this case (for the Python use case).